packages/devtools_app/lib/src/debugger/flutter/debugger_controller.dart

```diff
@@ -404,124 +420,55 @@ class DebuggerController extends DisposableController
     }
   }
 
+  Future<StackFrameAndSourcePosition> _createStackFrameWithLocation(
+    Frame frame,
+  ) {
+    final sf = StackFrameAndSourcePosition.create(frame);
+    return getScript(sf.script).then((Script script) {
+      final pos = calculatePosition(script, sf.tokenPos);
+      return StackFrameAndSourcePosition.create(frame, pos);
+    });
+  }
+
   void selectBreakpoint(BreakpointAndSourcePosition bp) {
     _selectedBreakpoint.value = bp;
   }
-}
-
-class SourcePosition {
-  SourcePosition({@required this.line, @required this.column, this.tokenPos});
 
-  final int line;
-  final int column;
-  final int tokenPos;
-
-  @override
-  String toString() => '$line:$column';
-}
-
-/// A tuple of a breakpoint and a source position.
-abstract class BreakpointAndSourcePosition
-    implements Comparable<BreakpointAndSourcePosition> {
-  BreakpointAndSourcePosition._(this.breakpoint, [this.sourcePosition]);
-
-  factory BreakpointAndSourcePosition.create(Breakpoint breakpoint,
-      [SourcePosition sourcePosition]) {
-    if (breakpoint.location is SourceLocation) {
-      return _BreakpointAndSourcePositionResolved(
-          breakpoint, sourcePosition, breakpoint.location as SourceLocation);
-    } else if (breakpoint.location is UnresolvedSourceLocation) {
-      return _BreakpointAndSourcePositionUnresolved(breakpoint, sourcePosition,
-          breakpoint.location as UnresolvedSourceLocation);
-    } else {
-      throw 'invalid value for breakpoint.location';
-    }
+  void selectStackFrame(StackFrameAndSourcePosition frame) {
+    _selectedStackFrame.value = frame;
   }
 
-  final Breakpoint breakpoint;
-  final SourcePosition sourcePosition;
-
-  bool get resolved => breakpoint.resolved;
-
-  ScriptRef get script;
-
-  String get scriptUri;
-
-  int get line;
-
-  int get column;
-
-  int get tokenPos;
-
-  String get id => breakpoint.id;
-
-  @override
-  int get hashCode => breakpoint.hashCode;
-
-  @override
-  bool operator ==(other) {
-    return other is BreakpointAndSourcePosition &&
-        other.breakpoint == breakpoint;
-  }
-
-  @override
-  int compareTo(BreakpointAndSourcePosition other) {
-    final result = scriptUri.compareTo(other.scriptUri);
-    if (result != 0) return result;
-
-    if (resolved != other.resolved) return resolved ? 1 : -1;
-
-    if (resolved) {
-      return tokenPos - other.tokenPos;
-    } else {
-      return line - other.line;
+  List<Frame> framesForCallStack() {
+    if (_callStack.value == null) return [];
+
+    List<Frame> frames =
+        _callStack.value.asyncCausalFrames ?? _callStack.value.frames;
+
+    // Handle breaking-on-exceptions.
+    if (_reportedException != null && frames.isNotEmpty) {
+      final Frame frame = frames.first;
+
+      final Frame newFrame = Frame(
```

Comment:
👍 this is great! You're creating a synthetic frame to hold the exception value?

We could also consider adding a new, synthetic variable to the top frame, in the 1st position in the list of variables.

Reply:
Ah, I didn't read the code closely enough; this does create a synthetic var in the first frame.